components/permission-provider/org.wso2.carbon.analytics.permissions/src/main/java/org/wso2/carbon/analytics/permissions/internal/impl/DefaultPermissionProvider.java

```diff
@@ -0,0 +1,251 @@
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
+package org.wso2.carbon.analytics.permissions.internal.impl;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.carbon.analytics.permissions.PermissionProvider;
+import org.wso2.carbon.analytics.permissions.bean.Permission;
+import org.wso2.carbon.analytics.permissions.bean.PermissionConfig;
+import org.wso2.carbon.analytics.permissions.bean.Role;
+import org.wso2.carbon.analytics.permissions.exceptions.PermissionException;
+import org.wso2.carbon.analytics.permissions.internal.dao.PermissionsDAO;
+import org.wso2.carbon.config.ConfigurationException;
+import org.wso2.carbon.config.provider.ConfigProvider;
+import org.wso2.carbon.datasource.core.api.DataSourceService;
+import org.wso2.carbon.stream.processor.idp.client.core.api.IdPClient;
+import org.wso2.carbon.stream.processor.idp.client.core.exception.IdPClientException;
+import org.wso2.carbon.stream.processor.idp.client.core.models.Group;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Default implementation of permission provider.
+ */
+@Component(
+        name = "DefaultPermissionProvider",
+        service = PermissionProvider.class,
+        immediate = true
+)
+public class DefaultPermissionProvider implements PermissionProvider {
+    private static final Logger log = LoggerFactory.getLogger(DefaultPermissionProvider.class);
+
+    private DataSourceService dataSourceService;
+    private IdPClient idPClient;
+    private PermissionConfig permissionConfig;
+
+    private PermissionsDAO permissionsDAO;
+
+    private PermissionsDAO getPermissionsDAO() {
+        if (permissionsDAO == null) {
+            log.debug("Permission DAO is not initialized. Initializing the DAO.");
+            permissionsDAO = new PermissionsDAO(dataSourceService, permissionConfig);
+        }
+        return permissionsDAO;
+    }
+
+    /**
+     * Add permission.
+     *
+     * @param permission
+     * @throws PermissionException
+     */
+    @Override
+    public void addPermission(Permission permission) throws PermissionException {
+        log.debug("Add permission " + permission);
```

Comment:
Let's add isDebugEnabled check since we are doing a string concatenation.

Reply:
Let's check other places as well.

Reply:
Ack.